pkg/logql/evaluator.go

```diff
@@ -547,14 +547,33 @@ func binOpStepEvaluator(
 		)
 	}
 
-	// we have two non literal legs
-	lse, err := ev.StepEvaluator(ctx, ev, expr.SampleExpr, q)
-	if err != nil {
-		return nil, err
-	}
-	rse, err := ev.StepEvaluator(ctx, ev, expr.RHS, q)
-	if err != nil {
-		return nil, err
+	// We have two non literal legs,
+	// load them in parallel
+	var (
+		lse, rse StepEvaluator
+		// prebuffer errCh so we can return after first error is found
+		errCh = make(chan error, 2)
+		// keep a scoped reference to err as it's referenced in the Error()
+		// implementation of this StepEvaluator
+		scopedErr error
+	)
+	go func() {
+		var err error
+		lse, err = ev.StepEvaluator(ctx, ev, expr.SampleExpr, q)
+		errCh <- err
+	}()
+	go func() {
+		var err error
+		rse, err = ev.StepEvaluator(ctx, ev, expr.RHS, q)
+		errCh <- err
+	}()
+
+	// ensure both sides are loaded before returning the combined evaluator
+	for i := 0; i < 2; i++ {
+		scopedErr = <-errCh
+		if scopedErr != nil {
+			return nil, scopedErr
+		}
```

Comment:
Could be simplified using `golang.org/x/sync/errgroup` ?

Reply:
I'm sure it could, thanks!